src/Stack/Upload.hs

```diff
@@ -90,12 +98,33 @@ loadCreds config = do
           "Save hackage credentials to file at " ++ fp ++ " [y/n]? "
         putStrLn "NOTE: Avoid this prompt in the future by using: save-hackage-creds: false"
         when shouldSave $ do
-          L.writeFile fp (encode hc)
+          writeFilePrivate fp $ fromEncoding $ toEncoding hc
           putStrLn "Saved!"
           hFlush stdout
 
       return hc
 
+-- | Write contents to a file which is always private.
+writeFilePrivate :: MonadIO m => FilePath -> Builder -> m ()
+writeFilePrivate fp builder = liftIO $ withTempFile (takeDirectory fp) (takeFileName fp) $ \fpTmp h0 -> do
+  -- Close the handle so we can set permissions first
+  hClose h0
+
+  -- Now set the permissions so only we can read and write
+  setFileMode fpTmp 0o600
+
+  -- Now append the contents we want
+  withBinaryFile fpTmp AppendMode $ \h -> hPutBuilder h builder
+
+  -- Disable writing
+  setFileMode fpTmp 0o400
+
+  -- Make sure the destination file, if present, is writeable
+  void $ tryIO $ setFileMode fp 0o600
+
+  -- And atomically move
+  renameFile fpTmp fp
+
```

Comment:
Is there a benefit in `-- Disable writing` on the temp file?

Also, it seems `setFileMode fpTmp 0o600` is unnecessary, as [`openTempFile`](https://www.stackage.org/haddock/lts-13.14/base-4.12.0.0/System-IO.html#v:openTempFile), which `withTempFile` uses, has in its docs:

> The file is created with permissions such that only the current user can read/write it.

Correspondingly, `hClose h0` should also be unnecessary.

Reply:
Oh, nice! I didn't notice that. Let me update.

Reply:
> Is there a benefit in -- Disable writing on the temp file?

I suppose not, I think I saw discussion of permissions 400 being set on the file, but I can't really think of a reason to use that. I'll remove that line too.